### ISSUE TITLE
Optimize the case of an async method with no await and remove the warning

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn#3659]
Currently, if I write a method like this:

```
public async Task<int> FooAsync()
{
    return 42;
}
```

The compiler emits the following warning:

> This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread

And if I look at the IL, I can see that the compiler actually emits a state machine, even tough none is needed.

I think this behavior could be improved.

- First, the compiler could just emit `return Task.FromResult(42)` (and do the same for every `return x`) if the method body contains no `await`. This would avoid the overhead of creating a state machine.

- Second, the warning is annoying. The fact that I use `await` or not in the body is just an implementation detail. If I change the implementation so that I no longer need `await`, I have to remove the `async` modifier and change every `return x` to `return Task.FromResult(x)` to avoid the warning. In some cases I could change the method signature to return `T` instead of `Task<T>`, but it's not always possible (e.g. if I'm implementing an interface or overriding a base class method). And if I change the implementation to make it asynchronous again, I have to do the opposite. IMO, the `async` modifier with no `await` should just mean "implicitly wrap any return value in a completed task". 

`async` without `await` should just be treated as a degenerate case, be optimized as such, and not emit a warning.

**Comment by HaloFour** [comment]
The state machine code also automatically wraps thrown exceptions.  A method marked as `async`, IIRC, can actually never throw an exception directly to the caller, it would always return a failed `Task<T>`.

**Comment by thomaslevesque** [comment]
@HaloFour, good point, I forgot about that. But the compiler could also emit the necessary code to wrap any exception in a failed task.

So it would transform this:

```
public async Task<int> FooAsync()
{
    return 42;
}
```

into this:

```
public Task<int> FooAsync()
{
    try
    {
        return Task.FromResult(42);
    }
    catch (Exception ex)
    {
        return Task.FromException<int>(ex);
    }
}
```

**Comment by HaloFour** [comment]
True.  I've not poked at the code for it but I bet the compiler starts to construct the state machine immediately on noticing that the method signature is `async` and only at the end realizes that there are no `await` statements but does not throw away and restart the work that it had done.

I think the warning exists more to prevent the case of the accidentally missed `await` statement.  In the previous C# project on which I worked I set that warning to be treated like an error as it occurred too often that a developer would forget to `await` on async methods returning `Task`.

**Comment by gafter** [comment]
See also #3815 for some correctness considerations.

**Comment by sami-daniel** [comment]
Hello everyone. Any chance this issue will continue to progress? It seems like something interesting to dedicate. 🫡

**Comment by CyrusNajmabadi** [comment]
@sami-daniel what's your need on this?  Also, this is not simple. As mentioned in linked issues, this would break some things like task-caching for certain types that the runtime performs.

IN terms of the warning, you can just disable that if you don't find it valuable.

**Comment by sami-daniel** [comment]
@CyrusNajmabadi I simply wanted to contribute to a stalled issue, but I'm not sure how complex it is (given the information provided above), as there are countless variables to consider beyond "returning `Task.FromResult(...)`."

Furthermore, I imagine it would be a very significant change, given that this behavior has been present in Roslyn for a long time. I could attempt this change as a research project or something like that.

**Comment by stbau04** [comment]
@CyrusNajmabadi IMHO it doesn't seem that complicated at all. I think it is more like a question of wether this actually brings any advantages. To avoid copying the things happnening in the AsyncTaskMethodBuilder into the generated code, either a new type would need to be defined or the `AsyncTaskMethodBuilder` would need to be used _(as the execution context cannot be set from external code as it is done within the `AsyncTaskMethodBuilder`, i think generating everything into the method wouldn't be working at all)_.

When using the `AsyncTaskMethodBuilder`, the state machine type would need to be emitted (as the execution context is only available internally). This would mean, that the code would exactly be the same _(maybe the state variable can be avoided - but that is only a very minor improvement i guess)_.

While the first approach would ensure, that the async always behaves exactly the same - regardless of whether awaits are used - the second approach would allow removing the state machine itself. Is optimizing this specific (probably even most of the time unwanted) situation actually worth introducing a new type to the runtime?

In any case i could not spot any task caching that is performed within the runtime on async without await. Sure that the there is task caching done in this situation?

**Comment by CyrusNajmabadi** [comment]
Closing out.  This will be optimized in the future with the new runtime async feature.  And we intend on removing the warning in that case.  Closing out as a dupe of that work: https://github.com/dotnet/roslyn/pull/80144
